File: sandbox/python/tracer.py
import sys
import json
import time
import traceback
import os
import builtins

# ─────────────────────────────────────────────────────────
#  Globals written to by the patched built-ins
# ─────────────────────────────────────────────────────────
_snapshots_out = sys.stdout   # always the real stdout
_start_time    = 0.0
_step_count    = 0
_prev_vars     = {}

def _ts():
    return (time.time() - _start_time) * 1000

def _emit(snapshot: dict):
    """Write a snapshot to the real stdout."""
    _snapshots_out.write("---SNAPSHOT---" + json.dumps(snapshot) + "\n")
    _snapshots_out.flush()

def _is_primitive(val):
    return isinstance(val, (int, float, str, bool, type(None)))

def _capture_heap_and_vars(frame):
    global _prev_vars
    variables = {}
    heap_map = {}

    def get_inspector_val(val):
        """Returns the user-friendly value to display in the Variable Inspector."""
        if _is_primitive(val):
            return val
        if isinstance(val, (list, tuple)):
            return [get_inspector_val(i) for i in val]
        if isinstance(val, set):
            return [get_inspector_val(i) for i in val]
        if isinstance(val, dict):
            return {str(k): get_inspector_val(v) for k, v in val.items()}
        # For custom objects, show Ref(TypeName) without the raw hex ID
        return f"Ref({type(val).__name__})"

    def process_heap(val, path_name=None):
        """Builds the heap map recursively."""
        if _is_primitive(val):
            return
        
        obj_id = f"0x{id(val):x}"
        type_name = type(val).__name__

        if obj_id in heap_map:
            if path_name and path_name not in heap_map[obj_id]["references"]:
                heap_map[obj_id]["references"].append(path_name)
            return

        heap_map[obj_id] = {
            "id": obj_id,
            "type": type_name,
            "value": None,
            "references": [path_name] if path_name else []
        }

        if isinstance(val, (list, tuple)):
            serialized = []
            for i, item in enumerate(val):
                process_heap(item, f"{path_name}[{i}]" if path_name else None)
                serialized.append(get_inspector_val(item))
            heap_map[obj_id]["value"] = serialized
        elif isinstance(val, set):
            serialized = []
            for item in val:
                process_heap(item, path_name if path_name else None)
                serialized.append(get_inspector_val(item))
            heap_map[obj_id]["value"] = list(serialized)
        elif isinstance(val, dict):
            serialized = {}
            for k, v in val.items():
                process_heap(v, f"{path_name}['{k}']" if path_name else None)
                serialized[str(k)] = get_inspector_val(v)
            heap_map[obj_id]["value"] = serialized
        else:
            serialized = {}
            if hasattr(val, '__dict__'):
                for k, v in val.__dict__.items():
                    if not k.startswith('__'):
                        process_heap(v, f"{path_name}.{k}" if path_name else None)
                        serialized[k] = get_inspector_val(v)
            else:
                serialized = str(val)
            heap_map[obj_id]["value"] = serialized

    # 1. Capture local variables
    for name, value in frame.f_locals.items():
        if name.startswith('__') or name.startswith('_'):
            continue
        # Exclude functions, classes, modules, and callables
        if callable(value) or type(value).__name__ in ('module', 'function', 'method', 'type', 'classobj', 'builtin_function_or_method'):
            continue
            
        process_heap(value, name)
        val_serialized = get_inspector_val(value)
        type_name = type(value).__name__
        changed = name not in _prev_vars or _prev_vars[name] != val_serialized
        variables[name] = {
            "name": name,
            "type": type_name,
            "value": val_serialized,
            "changed": changed,
            "scope": "local"
        }
        _prev_vars[name] = val_serialized

    # 2. Capture user globals
    for name, value in frame.f_globals.items():
        if name in ('__name__', '__doc__', '__package__', '__loader__', '__spec__', '__builtins__'):
            continue
        if name.startswith('_'):
            continue
        if name in variables:
            continue
        if callable(value) or type(value).__name__ in ('module', 'function', 'method', 'type', 'classobj', 'builtin_function_or_method'):
            continue
            
        process_heap(value, name)
        val_serialized = get_inspector_val(value)
        type_name = type(value).__name__
        changed = name not in _prev_vars or _prev_vars[name] != val_serialized
        variables[name] = {
            "name": name,
            "type": type_name,
            "value": val_serialized,
            "changed": changed,
            "scope": "global"
        }
        _prev_vars[name] = val_serialized

    return variables, list(heap_map.values())

def _capture_stack(frame):
    stack = []
    curr = frame
    while curr:
        if curr.f_code.co_filename.endswith('user_code.py'):
            frame_locals = {}
            for name, val in curr.f_locals.items():
                if name.startswith('__') or name.startswith('_'):
                    continue
                if callable(val) or type(val).__name__ in ('module', 'function', 'method', 'type', 'classobj', 'builtin_function_or_method'):
                    continue
                frame_locals[name] = str(val)
            stack.append({
                "function": curr.f_code.co_name,
                "line":     curr.f_lineno,
                "file":     os.path.basename(curr.f_code.co_filename),
                "locals":   frame_locals
            })
        curr = curr.f_back
    return stack

# ─────────────────────────────────────────────────────────
#  Patched input()
# ─────────────────────────────────────────────────────────
_real_input = builtins.input

def _patched_input(prompt=""):
    global _step_count
    sys.settrace(None)

    frame = sys._getframe(1)
    variables, heap = _capture_heap_and_vars(frame)
    stack     = _capture_stack(frame)
    line      = frame.f_lineno

    _emit({
        "step":         _step_count,
        "line":         line,
        "event":        "input_waiting",
        "label":        f"Program waiting for input{' (\"' + prompt + '\")' if prompt else ''}",
        "timestamp_ms": _ts(),
        "variables":    variables,
        "stack":        stack,
        "heap":         heap,
        "stdout_delta": prompt if prompt else "",
    })
    _step_count += 1

    value = _real_input("")

    variables2, heap2 = _capture_heap_and_vars(frame)

    _emit({
        "step":         _step_count,
        "line":         line,
        "event":        "input_received",
        "label":        f"Input received → \"{value}\"",
        "timestamp_ms": _ts(),
        "variables":    variables2,
        "stack":        stack,
        "heap":         heap2,
        "stdout_delta": "",
        "stdin_value":  value,
    })
    _step_count += 1

    sys.settrace(_trace_calls)
    return value

# ─────────────────────────────────────────────────────────
#  Patched print()
# ─────────────────────────────────────────────────────────
_real_print = builtins.print

def _patched_print(*args, sep=" ", end="\n", file=None, flush=False):
    global _step_count
    text = sep.join(str(a) for a in args) + end

    if file is not None and file is not sys.stdout:
        _real_print(*args, sep=sep, end=end, file=file, flush=flush)
        return

    sys.settrace(None)

    frame     = sys._getframe(1)
    variables, heap = _capture_heap_and_vars(frame)
    stack     = _capture_stack(frame)
    line      = frame.f_lineno

    _emit({
        "step":         _step_count,
        "line":         line,
        "event":        "output",
        "label":        f"Output → {text.rstrip()}",
        "timestamp_ms": _ts(),
        "variables":    variables,
        "stack":        stack,
        "heap":         heap,
        "stdout_delta": text,
    })
    _step_count += 1

    sys.settrace(_trace_calls)

# ─────────────────────────────────────────────────────────
#  sys.settrace callback
# ─────────────────────────────────────────────────────────
_max_steps = 1000

def _trace_calls(frame, event, arg):
    global _step_count

    if _step_count >= _max_steps:
        return None

    filename = frame.f_code.co_filename
    if not filename.endswith('user_code.py'):
        return _trace_calls

    if event not in ('line', 'call', 'return'):
        return _trace_calls

    variables, heap = _capture_heap_and_vars(frame)
    stack     = _capture_stack(frame)

    label = {
        'call':   f"Enter {frame.f_code.co_name}()",
        'return': f"Return from {frame.f_code.co_name}()",
        'line':   f"Execute line {frame.f_lineno}",
    }.get(event, f"Event: {event}")

    _emit({
        "step":         _step_count,
        "line":         frame.f_lineno,
        "event":        event,
        "label":        label,
        "timestamp_ms": _ts(),
        "variables":    variables,
        "stack":        stack,
        "heap":         heap,
        "stdout_delta": "",
    })
    _step_count += 1
    return _trace_calls

# ─────────────────────────────────────────────────────────
#  Entry point
# ─────────────────────────────────────────────────────────
def execute_user_code():
    global _start_time, _step_count, _prev_vars

    _start_time = time.time()
    _step_count = 0
    _prev_vars  = {}

    builtins.input = _patched_input
    builtins.print = _patched_print

    try:
        with open('user_code.py', 'r') as f:
            code = f.read()

        compiled_code = compile(code, 'user_code.py', 'exec')
        exec_globals  = {"__name__": "__main__"}

        sys.settrace(_trace_calls)
        exec(compiled_code, exec_globals)
        sys.settrace(None)

        final_vars = {}
        for name, val in _prev_vars.items():
            final_vars[name] = {
                "name": name,
                "type": "str" if isinstance(val, str) else type(val).__name__,
                "value": val,
                "changed": False,
                "scope": "local"
            }

        _emit({
            "step":         _step_count,
            "line":         None,
            "event":        "program_end",
            "label":        "Program finished",
            "timestamp_ms": _ts(),
            "variables":    final_vars,
            "stack":        [],
            "heap":         [],
            "stdout_delta": "",
        })
        _snapshots_out.flush()
        sys.stdout.flush()
        time.sleep(0.05)

    except Exception as e:
        sys.settrace(None)
        _emit({
            "step":         _step_count,
            "event":        "exception",
            "label":        f"Exception: {type(e).__name__}: {e}",
            "timestamp_ms": _ts(),
            "variables":    {},
            "stack":        [],
            "heap":         [],
            "stdout_delta": "",
            "exception": {
                "type":      type(e).__name__,
                "message":   str(e),
                "traceback": traceback.format_exc(),
            },
        })
    finally:
        builtins.input = _real_input
        builtins.print = _real_print
        sys.settrace(None)

if __name__ == "__main__":
    execute_user_code()

File: sandbox/java/Tracer.java
import com.sun.jdi.*;
import com.sun.jdi.connect.Connector;
import com.sun.jdi.connect.LaunchingConnector;
import com.sun.jdi.event.*;
import com.sun.jdi.request.*;

import java.io.*;
import java.util.*;

/**
 * Java Execution Tracer using JDI (Java Debug Interface).
 * Supports interactive stdin, real-time snapshot streaming, call-stack frame variables,
 * instance/static variables capture, and complete JDI Heap serialization.
 */
public class Tracer {
    private static final String TRACE_PREFIX = "---SNAPSHOT---";
    private static long startTime;
    private static int stepCount = 0;
    private static Map<String, Object> prevVars = new HashMap<>();
    
    // Step-local heap map to collect complex/reference object structures
    private static Map<String, Map<String, Object>> stepHeap = new LinkedHashMap<>();

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: java Tracer <ClassName>");
            System.exit(1);
        }

        String targetClassName = args[0];
        startTime = System.currentTimeMillis();
        VirtualMachine vm = null;

        try {
            VirtualMachineManager vmm = Bootstrap.virtualMachineManager();
            LaunchingConnector connector = vmm.defaultConnector();

            Map<String, Connector.Argument> arguments = connector.defaultArguments();
            arguments.get("main").setValue(targetClassName);
            
            // Build CP options ensuring the current directory is first for local compilation loading
            // IMPORTANT: JDI's LaunchingConnector passes options directly to the JVM process,
            // NOT through a shell. Do NOT use quotes around path components — they are treated
            // as literal characters and break classpath resolution.
            String classpath = System.getProperty("java.class.path");
            String cpOption;
            if (classpath != null && !classpath.isEmpty()) {
                cpOption = "-cp ." + File.pathSeparator + classpath;
            } else {
                cpOption = "-cp .";
            }
            arguments.get("options").setValue(cpOption);

            vm = connector.launch(arguments);
            final VirtualMachine finalVm = vm;

            // Forward Tracer's stdin to Target VM's stdin
            Thread stdinThread = new Thread(() -> {
                try (OutputStream vmStdin = finalVm.process().getOutputStream()) {
                    byte[] buffer = new byte[1024];
                    int n;
                    while ((n = System.in.read(buffer)) != -1) {
                        vmStdin.write(buffer, 0, n);
                        vmStdin.flush();
                    }
                } catch (IOException e) {}
            });
            stdinThread.setDaemon(true);
            stdinThread.start();

            // Redirect Target VM's stdout/stderr to Tracer's stdout
            Thread stdoutThread = new Thread(() -> redirectStream(finalVm.process().getInputStream(), false));
            stdoutThread.setDaemon(true);
            stdoutThread.start();

            Thread stderrThread = new Thread(() -> redirectStream(finalVm.process().getErrorStream(), true));
            stderrThread.setDaemon(true);
            stderrThread.start();

            EventRequestManager erm = vm.eventRequestManager();
            // We want to know when our target class is loaded
            erm.createClassPrepareRequest().enable();

            EventSet eventSet;
            boolean processRunning = true;

            while (processRunning && (eventSet = vm.eventQueue().remove()) != null) {
                for (Event event : eventSet) {
                    if (event instanceof VMStartEvent) {
                        // Initial thread is started and suspended
                    } else if (event instanceof ClassPrepareEvent) {
                        ClassPrepareEvent cpe = (ClassPrepareEvent) event;
                        if (cpe.referenceType().name().equals(targetClassName)) {
                            setupStepRequest(vm, cpe.thread());
                        }
                    } else if (event instanceof StepEvent) {
                        StepEvent stepEvent = (StepEvent) event;
                        
                        if (isInputLine(stepEvent)) {
                            emitInputWaiting(stepEvent);
                        } else {
                            captureSnapshot(stepEvent);
                        }
                    } else if (event instanceof VMDeathEvent || event instanceof VMDisconnectEvent) {
                        processRunning = false;
                    }
                }
                if (processRunning) {
                    vm.resume();
                }
            }

            // Wait slightly for output threads to finish reading remaining bytes
            try { Thread.sleep(300); } catch (InterruptedException e) {}

            // Emit a final snapshot to signal the end
            Map<String, Object> endSnapshot = new HashMap<>();
            endSnapshot.put("step", stepCount++);
            endSnapshot.put("event", "program_end");
            endSnapshot.put("label", "Program finished");
            endSnapshot.put("timestamp_ms", System.currentTimeMillis() - startTime);
            endSnapshot.put("variables", new HashMap<>());
            endSnapshot.put("stack", new ArrayList<>());
            endSnapshot.put("heap", new ArrayList<>());
            System.out.println(TRACE_PREFIX + toJson(endSnapshot));
            System.out.flush();
            
            // Final small delay to ensure stdout is flushed to parent
            try { Thread.sleep(100); } catch (InterruptedException e) {}

            System.exit(0);

        } catch (Exception e) {
            e.printStackTrace();
            if (vm != null) {
                try {
                    vm.exit(1);
                } catch (Exception ex) {}
            }
        }
    }

    private static void setupStepRequest(VirtualMachine vm, ThreadReference thread) {
        // Remove existing step requests for this thread if any
        vm.eventRequestManager().deleteEventRequests(vm.eventRequestManager().stepRequests());
        
        StepRequest sr = vm.eventRequestManager().createStepRequest(thread, StepRequest.STEP_LINE, StepRequest.STEP_INTO);
        // Exclude system classes instead of target class filter to allow debugging helper/inner classes
        sr.addClassExclusionFilter("java.*");
        sr.addClassExclusionFilter("javax.*");
        sr.addClassExclusionFilter("sun.*");
        sr.addClassExclusionFilter("com.sun.*");
        sr.addClassExclusionFilter("jdk.*");
        sr.addClassExclusionFilter("oracle.*");
        sr.enable();
    }

    private static boolean isInputLine(StepEvent event) {
        try {
            Location loc = event.location();
            int lineNum = loc.lineNumber();
            String sourceName = loc.sourceName();
            
            File sourceFile = new File(sourceName);
            if (sourceFile.exists()) {
                try (BufferedReader reader = new BufferedReader(new FileReader(sourceFile))) {
                    String lineText = "";
                    for (int i = 0; i < lineNum; i++) {
                        lineText = reader.readLine();
                    }
                    if (lineText != null) {
                        lineText = lineText.trim();
                        // Ignore comments
                        if (lineText.startsWith("//") || lineText.startsWith("/*") || lineText.startsWith("*")) {
                            return false;
                        }
                        // Match .next...() or .read...() but not 'new Scanner'
                        return (lineText.contains(".next") || lineText.contains(".read")) && 
                               !lineText.contains("new Scanner");
                    }
                }
            }
        } catch (Exception e) {}
        return false;
    }

    private static void emitInputWaiting(StepEvent event) {
        try {
            StackFrame frame = event.thread().frame(0);
            Location loc = frame.location();
            
            stepHeap.clear(); // Clear before capturing variables
            Map<String, Map<String, Object>> vars = captureVariables(frame);
            List<Map<String, Object>> heapList = new ArrayList<>(stepHeap.values());

            Map<String, Object> snapshot = new HashMap<>();
            snapshot.put("step", stepCount++);
            snapshot.put("line", loc.lineNumber());
            snapshot.put("event", "input_waiting");
            snapshot.put("label", "Program waiting for input...");
            snapshot.put("timestamp_ms", System.currentTimeMillis() - startTime);
            snapshot.put("variables", vars);
            snapshot.put("stack", captureStack(event.thread()));
            snapshot.put("heap", heapList);
            snapshot.put("stdout_delta", "");

            System.out.println(TRACE_PREFIX + toJson(snapshot));
            System.out.flush();
        } catch (Exception e) {}
    }

    private static Map<String, Map<String, Object>> captureVariables(StackFrame frame) {
        Map<String, Map<String, Object>> variables = new HashMap<>();
        try {
            // 1. Capture 'this' reference if in instance scope
            ObjectReference thisObj = frame.thisObject();
            if (thisObj != null) {
                String name = "this";
                processHeap(thisObj, name, 0, new HashSet<>());
                Object serializableValue = getInspectorValue(thisObj);
                String typeName = thisObj.referenceType().name();
                boolean changed = !serializableValue.equals(prevVars.get(name));
                
                Map<String, Object> varInfo = new HashMap<>();
                varInfo.put("name", name);
                varInfo.put("type", typeName);
                varInfo.put("value", serializableValue);
                varInfo.put("changed", changed);
                varInfo.put("scope", "local");
                
                variables.put(name, varInfo);
                prevVars.put(name, serializableValue);
            }

            // 2. Capture static fields (class level / global scope)
            ReferenceType refType = frame.location().declaringType();
            for (Field field : refType.allFields()) {
                if (field.isStatic()) {
                    String name = refType.name() + "." + field.name();
                    Value val = refType.getValue(field);
                    processHeap(val, name, 0, new HashSet<>());
                    Object serializableValue = getInspectorValue(val);
                    String typeName = field.typeName();
                    boolean changed = !serializableValue.equals(prevVars.get(name));
                    
                    Map<String, Object> varInfo = new HashMap<>();
                    varInfo.put("name", name);
                    varInfo.put("type", typeName);
                    varInfo.put("value", serializableValue);
                    varInfo.put("changed", changed);
                    varInfo.put("scope", "global");
                    
                    variables.put(name, varInfo);
                    prevVars.put(name, serializableValue);
                }
            }

            // 3. Capture local variables
            for (LocalVariable var : frame.visibleVariables()) {
                Value val = frame.getValue(var);
                processHeap(val, var.name(), 0, new HashSet<>());
                Object serializableValue = getInspectorValue(val);
                String typeName = var.typeName();
                boolean changed = !serializableValue.equals(prevVars.get(var.name()));
                
                Map<String, Object> varInfo = new HashMap<>();
                varInfo.put("name", var.name());
                varInfo.put("type", typeName);
                varInfo.put("value", serializableValue);
                varInfo.put("changed", changed);
                varInfo.put("scope", "local");
                
                variables.put(var.name(), varInfo);
                prevVars.put(var.name(), serializableValue);
            }
        } catch (Exception e) {}
        return variables;
    }

    private static Object getInspectorValue(Value val) {
        if (val == null) return null;
        if (val instanceof StringReference) return ((StringReference) val).value();
        if (val instanceof PrimitiveValue) return val.toString();
        if (val instanceof ArrayReference) {
            ArrayReference arr = (ArrayReference) val;
            List<Object> list = new ArrayList<>();
            try {
                for (int i = 0; i < arr.length(); i++) {
                    list.add(getInspectorValue(arr.getValue(i)));
                }
            } catch (Exception e) {}
            return list;
        }
        if (val instanceof ObjectReference) {
            ObjectReference obj = (ObjectReference) val;
            String typeName = obj.referenceType().name();
            // For JDK system/common classes, just return their toString representation
            if (typeName.startsWith("java.lang.String")) {
                return obj.toString();
            }
            if (typeName.startsWith("java.") || typeName.startsWith("javax.") || typeName.startsWith("sun.")) {
                return obj.toString();
            }
            // For custom objects, show Ref(TypeName) without the unique address ID to prevent clutter
            return "Ref(" + typeName + ")";
        }
        return val.toString();
    }

    private static String getSimpleStringValue(Value val) {
        if (val == null) return "null";
        if (val instanceof StringReference) return "\"" + ((StringReference) val).value() + "\"";
        if (val instanceof PrimitiveValue) return val.toString();
        if (val instanceof ArrayReference) {
            return "Array(len=" + ((ArrayReference) val).length() + ")";
        }
        if (val instanceof ObjectReference) {
            return "Object(" + ((ObjectReference) val).referenceType().name() + ")";
        }
        return val.toString();
    }

    private static List<Map<String, Object>> captureStack(ThreadReference thread) {
        List<Map<String, Object>> stack = new ArrayList<>();
        try {
            for (int i = 0; i < thread.frameCount(); i++) {
                StackFrame f = thread.frame(i);
                Location l = f.location();
                Map<String, Object> stackItem = new HashMap<>();
                stackItem.put("function", l.method().name());
                stackItem.put("line", l.lineNumber());
                try {
                    stackItem.put("file", l.sourceName());
                } catch (Exception e) {
                    stackItem.put("file", "unknown");
                }
                
                // Extract frame locals to support StackVisualizer.jsx
                Map<String, Object> locals = new HashMap<>();
                try {
                    for (LocalVariable var : f.visibleVariables()) {
                        Value val = f.getValue(var);
                        locals.put(var.name(), getSimpleStringValue(val));
                    }
                } catch (Exception e) {}
                stackItem.put("locals", locals);
                
                stack.add(stackItem);
            }
        } catch (Exception e) {}
        return stack;
    }

    private static void captureSnapshot(StepEvent event) {
        try {
            StackFrame frame = event.thread().frame(0);
            Location loc = frame.location();

            stepHeap.clear(); // Clear before capturing variables
            Map<String, Map<String, Object>> vars = captureVariables(frame);
            List<Map<String, Object>> heapList = new ArrayList<>(stepHeap.values());

            Map<String, Object> snapshot = new HashMap<>();
            snapshot.put("step", stepCount++);
            snapshot.put("line", loc.lineNumber());
            snapshot.put("event", "line");
            snapshot.put("label", "Execute line " + loc.lineNumber());
            snapshot.put("timestamp_ms", System.currentTimeMillis() - startTime);
            snapshot.put("variables", vars);
            snapshot.put("stack", captureStack(event.thread()));
            snapshot.put("heap", heapList);
            snapshot.put("stdout_delta", "");

            System.out.println(TRACE_PREFIX + toJson(snapshot));
            System.out.flush();

        } catch (Exception e) {}
    }

    @SuppressWarnings("unchecked")
    private static void processHeap(Value val, String pathName, int depth, Set<Long> visited) {
        if (val == null || depth > 5) return;

        if (val instanceof ArrayReference) {
            ArrayReference arr = (ArrayReference) val;
            String refId = "0x" + Long.toHexString(arr.uniqueID());
            
            // Add or update heap entry
            Map<String, Object> heapEntry = stepHeap.get(refId);
            boolean isNew = false;
            if (heapEntry == null) {
                heapEntry = new LinkedHashMap<>();
                heapEntry.put("id", refId);
                heapEntry.put("type", arr.referenceType().name());
                heapEntry.put("references", new ArrayList<String>());
                stepHeap.put(refId, heapEntry);
                isNew = true;
            }
            if (pathName != null) {
                List<String> refs = (List<String>) heapEntry.get("references");
                if (!refs.contains(pathName)) {
                    refs.add(pathName);
                }
            }
            
            if (isNew) {
                List<Object> list = new ArrayList<>();
                try {
                    for (int i = 0; i < arr.length(); i++) {
                        Value item = arr.getValue(i);
                        String subPath = (pathName != null) ? pathName + "[" + i + "]" : null;
                        processHeap(item, subPath, depth + 1, visited);
                        list.add(getInspectorValue(item));
                    }
                } catch (Exception e) {}
                heapEntry.put("value", list);
            }
        }

        if (val instanceof ObjectReference) {
            ObjectReference obj = (ObjectReference) val;
            long id = obj.uniqueID();
            String refId = "0x" + Long.toHexString(id);
            
            // Add or update heap entry
            Map<String, Object> heapEntry = stepHeap.get(refId);
            boolean isNew = false;
            if (heapEntry == null) {
                heapEntry = new LinkedHashMap<>();
                heapEntry.put("id", refId);
                heapEntry.put("type", obj.referenceType().name());
                heapEntry.put("references", new ArrayList<String>());
                stepHeap.put(refId, heapEntry);
                isNew = true;
            }
            if (pathName != null) {
                List<String> refs = (List<String>) heapEntry.get("references");
                if (!refs.contains(pathName)) {
                    refs.add(pathName);
                }
            }

            if (isNew) {
                if (visited.contains(id)) {
                    heapEntry.put("value", "Ref(id=" + id + ")");
                } else {
                    visited.add(id);
                    ReferenceType type = obj.referenceType();
                    String typeName = type.name();
                    
                    // For JDK system/common classes, just return their toString representation
                    if (typeName.startsWith("java.lang.String")) {
                        heapEntry.put("value", obj.toString());
                    } else if (typeName.startsWith("java.") || typeName.startsWith("javax.") || typeName.startsWith("sun.")) {
                        heapEntry.put("value", obj.toString());
                    } else {
                        Map<String, Object> fields = new LinkedHashMap<>();
                        try {
                            for (Field field : type.allFields()) {
                                if (!field.isStatic()) {
                                    String subPath = (pathName != null) ? pathName + "." + field.name() : null;
                                    processHeap(obj.getValue(field), subPath, depth + 1, visited);
                                    fields.put(field.name(), getInspectorValue(obj.getValue(field)));
                                }
                            }
                            if (fields.isEmpty()) {
                                heapEntry.put("value", obj.toString());
                            } else {
                                heapEntry.put("value", fields);
                            }
                        } catch (Exception e) {
                            heapEntry.put("value", obj.toString());
                        }
                    }
                    visited.remove(id);
                }
            }
        }
    }

    private static void redirectStream(InputStream in, boolean isError) {
        try {
            InputStreamReader reader = new InputStreamReader(in);
            char[] buffer = new char[1024];
            int n;
            while ((n = reader.read(buffer)) != -1) {
                String text = new String(buffer, 0, n);
                Map<String, Object> snapshot = new HashMap<>();
                snapshot.put("step", stepCount++);
                snapshot.put("event", "output");
                snapshot.put("label", "Output: " + text.trim());
                snapshot.put("stdout_delta", text);
                snapshot.put("timestamp_ms", System.currentTimeMillis() - startTime);
                snapshot.put("variables", new HashMap<>());
                snapshot.put("stack", new ArrayList<>());
                snapshot.put("heap", new ArrayList<>());
                
                System.out.println(TRACE_PREFIX + toJson(snapshot));
                System.out.flush();
            }
        } catch (IOException e) {}
    }

    private static String toJson(Object obj) {
        if (obj == null) return "null";
        if (obj instanceof String) return "\"" + escape((String) obj) + "\"";
        if (obj instanceof Number || obj instanceof Boolean) return obj.toString();
        if (obj instanceof Map) {
            StringBuilder sb = new StringBuilder("{");
            Map<?, ?> map = (Map<?, ?>) obj;
            boolean first = true;
            for (Map.Entry<?, ?> entry : map.entrySet()) {
                if (!first) sb.append(",");
                sb.append("\"").append(entry.getKey()).append("\":").append(toJson(entry.getValue()));
                first = false;
            }
            sb.append("}");
            return sb.toString();
        }
        if (obj instanceof List) {
            StringBuilder sb = new StringBuilder("[");
            List<?> list = (List<?>) obj;
            boolean first = true;
            for (Object item : list) {
                if (!first) sb.append(",");
                sb.append(toJson(item));
                first = false;
            }
            sb.append("]");
            return sb.toString();
        }
        return "\"" + escape(obj.toString()) + "\"";
    }

    private static String escape(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r");
    }
}
